kong/tools/utils.lua

```diff
@@ -1295,6 +1298,74 @@ do
 ]],
   }
 
+  local ngx_log = ngx.log
+  local ERR     = ngx.ERR
+  local custom_error_templates = setmetatable({}, {
+    __index = function(self, format)
+      local template_path = kong.configuration["error_template_" .. format]
+      if not template_path then
+        rawset(self, format, false)
+        return false
+      end
+
+      local template, err
+      if pl_path.exists(template_path) then
+        template, err = pl_file.read(template_path)
+      else
+        err = "file not found"
+      end
+
+      if template then
+        rawset(self, format, template)
+        return template
+      end
+
+      ngx_log(ERR, fmt("failed reading the custom %s error template: %s", format, err))
+      rawset(self, format, false)
+      return false
+    end
+  })
+
+
+  get_response_type = function(accept_header)
+    local content_type = MIME_TYPES[CONTENT_TYPE_DEFAULT]
+    if type(accept_header) == "table" then
+      accept_header = join(",", accept_header)
+    end
+
+    if accept_header ~= nil then
+      local token_pattern = "[a-z0-9][a-z0-9-!#$&^_+.]+"
+      local pattern = string.format([[
+        ((?:%s|\*) \/ (?:%s|\*))
+        (?:
+          \s*;\s*
+          q = ( 1(?:\.0{0,3}|) | 0(?:\.\d{0,3}|) )
+          | \s*;\s* %s (?:=[^;]*|)
+        )*
+      ]], token_pattern, token_pattern, token_pattern)
```

Comment:
I'm not a huge fan of putting the token pattern into a variable and then interpolating because it mixes two languages, making it overall harder to understand what's going on.  If you must do it, then I'd recommend doing it like this:
```
local pattern = string.gsub([[
        ((?:TOKEN|\*) \/ (?:TOKEN|\*))
        (?:
          \s*;\s*
          q = ( 1(?:\.0{0,3}|) | 0(?:\.\d{0,3}|) )
          | \s*;\s* TOKEN (?:=[^;]*|)
        )*
      ]], "TOKEN", "[a-z0-9][a-z0-9-!#$&^_+.]+")
```
That way, it is easier to distinguish the TOKEN replacement string from `\s`, and it also does not require counting.  But I'd just leave the token pattern inside of the main regular expression.

In any case, you want to push the possible assembly and compilation of the regular expression into the global scope as there is no need to do it for every invocation of `get_response_type`, and the performance gang would immediately complain.

Reply:
> you want to push the possible assembly and compilation of the regular expression into the global scope as there is no need to do it for every invocation of get_response_type

since we are using the option `o` we should be ok I believe: _the regex pattern in question will only be compiled once, cached in the worker-process level, and shared among all requests in the current Nginx worker process_ see: https://github.com/openresty/lua-nginx-module#ngxrematch